File: backend/app/routers/agent_memory.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.deps import get_agent_or_404, get_agent_owner_or_404
from app.models.agent import Agent
from app.schemas.agent_memory import AgentMemoryCreate, AgentMemoryOut
from app.services import agent_memory_service

router = APIRouter(prefix="/agents/{agent_id}/memory", tags=["agent-memory"])


@router.get("", response_model=list[AgentMemoryOut])
def list_memory(agent: Agent = Depends(get_agent_or_404), db: Session = Depends(get_db)) -> list:
    return agent_memory_service.list_memories(db, agent.id)


@router.post("", response_model=AgentMemoryOut)
def create_memory(
    payload: AgentMemoryCreate,
    agent: Agent = Depends(get_agent_owner_or_404),
    db: Session = Depends(get_db),
) -> AgentMemoryOut:
    return agent_memory_service.create_memory(db, agent.id, payload.content)


@router.delete("/{memory_id}")
def delete_memory(
    memory_id: str,
    agent: Agent = Depends(get_agent_owner_or_404),
    db: Session = Depends(get_db),
) -> dict:
    memory = agent_memory_service.get_memory_or_none(db, agent.id, memory_id)
    if not memory:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Business rule not found")
    agent_memory_service.delete_memory(db, memory)
    return {"ok": True}

File: frontend/src/api/agentMemory.ts
import { apiClient } from "./client";

export interface AgentMemory {
  id: string;
  agent_id: string;
  content: string;
  created_at: string;
  updated_at: string;
}

export async function listAgentMemory(agentId: string): Promise<AgentMemory[]> {
  const { data } = await apiClient.get<AgentMemory[]>(`/agents/${agentId}/memory`);
  return data;
}

export async function createAgentMemory(agentId: string, content: string): Promise<AgentMemory> {
  const { data } = await apiClient.post<AgentMemory>(`/agents/${agentId}/memory`, { content });
  return data;
}

export async function deleteAgentMemory(agentId: string, memoryId: string): Promise<void> {
  await apiClient.delete(`/agents/${agentId}/memory/${memoryId}`);
}
